external/kafka-0-10-sql/src/test/scala/org/apache/spark/sql/kafka010/InternalKafkaConsumerPoolSuite.scala

```diff
@@ -0,0 +1,316 @@
+
+import org.apache.kafka.clients.consumer.ConsumerConfig._
+import org.apache.kafka.common.TopicPartition
+import org.apache.kafka.common.serialization.ByteArrayDeserializer
+
+import org.apache.spark.SparkEnv
+import org.apache.spark.sql.kafka010.KafkaDataConsumer.CacheKey
+import org.apache.spark.sql.test.SharedSQLContext
+
+class InternalKafkaConsumerPoolSuite extends SharedSQLContext {
+  import org.apache.spark.sql.kafka010.InternalKafkaConsumerPool.PoolConfig._
+
+  test("basic multiple borrows and returns for single key") {
+    val pool = InternalKafkaConsumerPool.build
+
+    val topic = "topic"
+    val partitionId = 0
+    val topicPartition = new TopicPartition(topic, partitionId)
+
+    val kafkaParams: ju.Map[String, Object] = getTestKafkaParams
+
+    val key = new CacheKey(topicPartition, kafkaParams)
+
+    val pooledObjects = (0 to 2).map { _ =>
+      val pooledObject = pool.borrowObject(key, kafkaParams)
+      assertPooledObject(pooledObject, topicPartition, kafkaParams)
+      pooledObject
+    }
+
+    assertPoolStateForKey(pool, key, numIdle = 0, numActive = 3, numTotal = 3)
+    assertPoolState(pool, numIdle = 0, numActive = 3, numTotal = 3)
+
+    val pooledObject2 = pool.borrowObject(key, kafkaParams)
+
+    assertPooledObject(pooledObject2, topicPartition, kafkaParams)
+    assertPoolStateForKey(pool, key, numIdle = 0, numActive = 4, numTotal = 4)
+    assertPoolState(pool, numIdle = 0, numActive = 4, numTotal = 4)
+
+    pooledObjects.foreach(pool.returnObject)
+
+    assertPoolStateForKey(pool, key, numIdle = 3, numActive = 1, numTotal = 4)
+    assertPoolState(pool, numIdle = 3, numActive = 1, numTotal = 4)
+
+    pool.returnObject(pooledObject2)
+
+    // we only allow three idle objects per key
+    assertPoolStateForKey(pool, key, numIdle = 3, numActive = 0, numTotal = 3)
+    assertPoolState(pool, numIdle = 3, numActive = 0, numTotal = 3)
+
+    pool.close()
+  }
+
+  test("basic borrow and return for multiple keys") {
+    val pool = InternalKafkaConsumerPool.build
+
+    val kafkaParams = getTestKafkaParams
+    val topicPartitions: List[TopicPartition] = for (
+      topic <- List("topic", "topic2");
+      partitionId <- 0 to 5
+    ) yield new TopicPartition(topic, partitionId)
+
+    val keys: List[CacheKey] = topicPartitions.map { part =>
+      new CacheKey(part, kafkaParams)
+    }
+
+    // while in loop pool doesn't still exceed total pool size
+    val keyToPooledObjectPairs = borrowObjectsPerKey(pool, kafkaParams, keys)
+
+    assertPoolState(pool, numIdle = 0, numActive = keyToPooledObjectPairs.length,
+      numTotal = keyToPooledObjectPairs.length)
+
+    returnObjects(pool, keyToPooledObjectPairs)
+
+    assertPoolState(pool, numIdle = keyToPooledObjectPairs.length, numActive = 0,
+      numTotal = keyToPooledObjectPairs.length)
+
+    pool.close()
+  }
+
+  test("borrow more than soft max capacity from pool which is neither free space nor idle object") {
```

Comment:
I have the feeling `borrow...` tests have similarities.

Reply:
I'll revisit and see we can deduplicate something. It seems to have some duplications between tests as you said.

Reply:
Addressed.